src/lib.rs

```diff
@@ -46,16 +46,280 @@ pub enum JsMonth {
+    match self {
+      RRuleType::Unvalidated(rrule) => rrule.get_week_start(),
+      RRuleType::Validated(rrule) => rrule.get_week_start(),
+    }
+  }
+
+  pub fn get_until(&self) -> Option<&DateTime<Tz>> {
+    match self {
+      RRuleType::Unvalidated(rrule) => rrule.get_until(),
+      RRuleType::Validated(rrule) => rrule.get_until(),
+    }
+  }
+
+  pub fn interval(self, interval: u16) -> Self {
+    match self {
+      RRuleType::Unvalidated(rrule) => RRuleType::Unvalidated(rrule.interval(interval)),
+      RRuleType::Validated(rrule) => {
+        RRuleType::Unvalidated(to_unvalidated(&rrule).interval(interval))
+      }
+    }
+  }
+
+  pub fn count(self, count: u32) -> Self {
+    match self {
+      RRuleType::Unvalidated(rrule) => RRuleType::Unvalidated(rrule.count(count)),
+      RRuleType::Validated(rrule) => RRuleType::Unvalidated(to_unvalidated(&rrule).count(count)),
+    }
+  }
+
+  pub fn by_weekday(self, weekdays: Vec<NWeekday>) -> Self {
+    match self {
+      RRuleType::Unvalidated(rrule) => RRuleType::Unvalidated(rrule.by_weekday(weekdays)),
+      RRuleType::Validated(rrule) => {
+        RRuleType::Unvalidated(to_unvalidated(&rrule).by_weekday(weekdays))
+      }
+    }
+  }
+
+  pub fn by_hour(self, hours: Vec<u8>) -> Self {
+    match self {
+      RRuleType::Unvalidated(rrule) => RRuleType::Unvalidated(rrule.by_hour(hours)),
+      RRuleType::Validated(rrule) => RRuleType::Unvalidated(to_unvalidated(&rrule).by_hour(hours)),
+    }
+  }
+
+  pub fn by_minute(self, minutes: Vec<u8>) -> Self {
+    match self {
+      RRuleType::Unvalidated(rrule) => RRuleType::Unvalidated(rrule.by_minute(minutes)),
+      RRuleType::Validated(rrule) => {
+        RRuleType::Unvalidated(to_unvalidated(&rrule).by_minute(minutes))
+      }
+    }
+  }
+
+  pub fn by_second(self, seconds: Vec<u8>) -> Self {
+    match self {
+      RRuleType::Unvalidated(rrule) => RRuleType::Unvalidated(rrule.by_second(seconds)),
+      RRuleType::Validated(rrule) => {
+        RRuleType::Unvalidated(to_unvalidated(&rrule).by_second(seconds))
+      }
+    }
+  }
+
+  pub fn by_month_day(self, days: Vec<i8>) -> Self {
+    match self {
+      RRuleType::Unvalidated(rrule) => RRuleType::Unvalidated(rrule.by_month_day(days)),
+      RRuleType::Validated(rrule) => {
+        RRuleType::Unvalidated(to_unvalidated(&rrule).by_month_day(days))
+      }
+    }
+  }
+
+  pub fn by_set_pos(self, poses: Vec<i32>) -> Self {
+    match self {
+      RRuleType::Unvalidated(rrule) => RRuleType::Unvalidated(rrule.by_set_pos(poses)),
+      RRuleType::Validated(rrule) => {
+        RRuleType::Unvalidated(to_unvalidated(&rrule).by_set_pos(poses))
+      }
+    }
+  }
+
+  pub fn by_month(self, months: &[Month]) -> Self {
+    match self {
+      RRuleType::Unvalidated(rrule) => RRuleType::Unvalidated(rrule.by_month(months)),
+      RRuleType::Validated(rrule) => {
+        RRuleType::Unvalidated(to_unvalidated(&rrule).by_month(months))
```

Comment:
It looks to much to create a new object on every property change. Also I see many code duplications in all setters and getters.

I suggest you to create unvalidated object only in getRrules methods:

```rust
  #[napi(ts_return_type="RRule[]")]
  pub fn get_rrules(&self, env: Env) -> napi::Result<Array> {
    let mut arr = env.create_array(0).unwrap();
    let rrules = self.rrule_set.get_rrule();

    for rrule in rrules.iter() {
      arr.insert(JsRRule {
        rrule: to_unvalidated(rrule)
      }).unwrap();
    }

    Ok(arr)
  }

```

This way, modifications of `RRule` object received from `getRrules` method won't be that expensive, regular `RRule` and `RRule` objects from `getRrules` will behave the same, plus it will significantly shrink amount of new code which should be covered by tests.

The same relates to `get_exrules` method

Reply:
The idea was to have a copy-on-write behavior instead of creating copies on read but I agree that the amount of code needed for this doesn't justify the benefits.